traits/trait_list_object.py

```diff
@@ -53,460 +94,754 @@ def __init__(self, index=0, removed=None, added=None):
-                    # are.
-                    index = 0 if key.start is None else key.start
-                else:
-                    index = key
-            else:
-                if validate is not None:
-                    value = validate(object, name, value)
-
-                values = [value]
-                removed = [removed]
-                delta = 0
-
-                index = len(self) + key if key < 0 else key
-
-            list.__setitem__(self, key, value)
-            if self.name_items is not None:
-                if delta == 0:
-                    try:
-                        if removed == values:
-                            return
-                    except Exception:
-                        # Treat incomparable values as equal:
-                        pass
-                self._send_trait_items_event(
-                    self.name_items, TraitListEvent(index, removed, values)
-                )
+    def __getstate__(self):
+        """ Get the state of the object for serialization.
 
-        except TraitError as excp:
-            excp.set_prefix("Each element of the")
-            raise excp
+        Notifiers are transient and should not be serialized.
+        """
+        result = self.__dict__.copy()
+        result.pop("notifiers", None)
+        return result
 
-    def __delitem__(self, key):
-        trait = getattr(self, "trait", None)
-        if trait is None:
-            return list.__delitem__(self, key)
+    def __setstate__(self, state):
+        """ Restore the state of the object after serialization.
 
-        try:
-            removed = self[key]
-        except Exception:
-            removed = []
+        Notifiers are transient and are restored to the empty list.
+        """
+        state['notifiers'] = []
+        self.__dict__.update(state)
 
-        if isinstance(key, slice):
-            slice_len = len(removed)
-            delta = slice_len
-            step = 1 if key.step is None else key.step
-            if step == 1:
-                # FIXME: See corresponding comment in __setitem__() for
-                # explanation.
-                index = 0 if key.start is None else key.start
-            else:
-                index = key
+    def __setitem__(self, index, value):
+        """ Set a value at index, implements self[index] = value
+
+        Parameters
+        ----------
+        index : int
+            Index at which the value will be set.
+        value : Any
+            The value to set at the index.
+
+        Notes
+        -----
+        Notification:
+            index : int
+                The index of the value.
+            added : list
+                The value that is set, as a list.
+            removed : list
+                The value that is removed, as a list.
+
+        """
+        removed = self._get_removed(index, default=[])
+        is_index_slice = isinstance(index, slice)
```

Comment:
nit: how about `index_is_slice` instead of `is_index_slice`? My brain has to think to parse `is_index_slice`.